File: 04.2_diffie_hellman/diffie_hellman.cpp
// To compile and run this code, open a terminal in this folder and run:
// g++ diffie_hellman.cpp -o diffie_hellman && ./diffie_hellman

#include <iostream>
#include <string>

// Use long long for wider integer range, suitable for this example's numbers.
using ll = long long;

/**
 * @brief Power function to return value of (base^exp) mod modulus.
 * * @param base The base of the exponentiation.
 * @param exp The exponent.
 * @param modulus The modulus.
 * @return The result of (base^exp) mod modulus.
 */
ll power(ll base, ll exp, ll modulus) {
    ll result = 1;
    base %= modulus;
    while (exp > 0) {
        if (exp % 2 == 1) {
            result = (result * base) % modulus;
        }
        exp >>= 1; // equivalent to exp = exp / 2
        base = (base * base) % modulus;
    }
    return result;
}

/**
 * @brief A simple XOR-based encryption/decryption function.
 * * Applying the same function twice with the same key returns the original message.
 * @param message The string message to be processed.
 * @param key The secret key to use for the XOR operation.
 * @return The processed (encrypted or decrypted) string.
 */
std::string encryptDecrypt(const std::string& message, ll key) {
    std::string result = "";
    for (char c : message) {
        result += c ^ key;
    }
    return result;
}

int main() {
    ll P, G, x, a, y, b, ka, kb;

    // Both parties agree upon the public keys G and P.

    // A prime number P is taken.
    P = 23;
    std::cout << "The value of P: " << P << std::endl;

    // A primitive root for P, G is taken.
    G = 9;
    std::cout << "The value of G: " << G << std::endl;
    std::cout << "-------------------------" << std::endl;

    // Alice chooses her private key 'a'.
    a = 4;
    std::cout << "The private key 'a' for Alice: " << a << std::endl;

    // Alice calculates her public key 'x'.
    x = power(G, a, P);
    std::cout << "The public key 'x' for Alice: " << x << std::endl;
    std::cout << "-------------------------" << std::endl;

    // Bob chooses his private key 'b'.
    b = 3;
    std::cout << "The private key 'b' for Bob: " << b << std::endl;

    // Bob calculates his public key 'y'.
    y = power(G, b, P);
    std::cout << "The public key 'y' for Bob: " << y << std::endl;
    std::cout << "-------------------------" << std::endl;

    // After exchanging public keys, they generate the shared secret key.

    // Alice calculates the secret key.
    ka = power(y, a, P);

    // Bob calculates the secret key.
    kb = power(x, b, P);

    std::cout << "Secret key for Alice is: " << ka << std::endl;
    std::cout << "Secret key for Bob is: " << kb << std::endl;
    std::cout << "-------------------------" << std::endl;

    // Now they can use the shared secret key for encryption.
    std::string message = "Hello Bob!";
    std::cout << "Original Message: " << message << std::endl;

    // Alice encrypts the message with her key.
    std::string encryptedMessage = encryptDecrypt(message, ka);
    std::cout << "Encrypted Message: " << encryptedMessage << std::endl;

    // Bob decrypts the message with his key.
    std::string decryptedMessage = encryptDecrypt(encryptedMessage, kb);
    std::cout << "Decrypted Message: " << decryptedMessage << std::endl;

    return 0;
}

File: 04.2_diffie_hellman/diffie_hellman.js
// To run this code, you need Node.js installed.
// Open a terminal in this folder and run: node diffie_hellman.js

/**
 * Power function to return value of (a^b) mod P.
 * This uses BigInt for handling potentially large numbers in cryptography.
 *
 * @param {bigint} a The base.
 * @param {bigint} b The exponent.
 * @param {bigint} P The modulus.
 * @returns {bigint} The result of (a^b) mod P.
 */
function power(a, b, P) {
    // Modular exponentiation
    let result = 1n;
    a = a % P;
    while (b > 0n) {
        if (b % 2n === 1n) {
            result = (result * a) % P;
        }
        b = b / 2n;
        a = (a * a) % P;
    }
    return result;
}

/**
 * A simple XOR-based encryption/decryption function.
 * Applying the same function twice with the same key returns the original message.
 *
 * @param {string} message The string message to be processed.
 * @param {bigint} key The secret key to use for the XOR operation.
 * @returns {string} The processed (encrypted or decrypted) string.
 */
function encryptDecrypt(message, key) {
    let result = '';
    // The key might be a BigInt, so we convert it to a Number for XOR.
    // This is safe for typical key sizes in this example.
    const keyNum = Number(key);
    for (let i = 0; i < message.length; i++) {
        result += String.fromCharCode(message.charCodeAt(i) ^ keyNum);
    }
    return result;
}

function main() {
    // We use 'n' at the end of numbers to denote them as BigInt literals.
    let P, G, x, a, y, b, ka, kb;

    // Both parties agree upon the public keys G and P.

    // A prime number P is taken.
    P = 23n;
    console.log(`The value of P: ${P}`);

    // A primitive root for P, G is taken.
    G = 9n;
    console.log(`The value of G: ${G}`);
    console.log("-------------------------");

    // Alice chooses her private key 'a'.
    a = 4n;
    console.log(`The private key 'a' for Alice: ${a}`);

    // Alice calculates her public key 'x'.
    x = power(G, a, P);
    console.log(`The public key 'x' for Alice: ${x}`);
    console.log("-------------------------");

    // Bob chooses his private key 'b'.
    b = 3n;
    console.log(`The private key 'b' for Bob: ${b}`);

    // Bob calculates his public key 'y'.
    y = power(G, b, P);
    console.log(`The public key 'y' for Bob: ${y}`);
    console.log("-------------------------");

    // After exchanging public keys, they generate the shared secret key.

    // Alice calculates the secret key.
    ka = power(y, a, P);

    // Bob calculates the secret key.
    kb = power(x, b, P);

    console.log(`Secret key for Alice is: ${ka}`);
    console.log(`Secret key for Bob is: ${kb}`);
    console.log("-------------------------");

    // Now they can use the shared secret key for encryption.
    const message = "Hello Bob!";
    console.log(`Original Message: ${message}`);

    // Alice encrypts the message with her key.
    const encryptedMessage = encryptDecrypt(message, ka);
    console.log(`Encrypted Message: ${encryptedMessage}`);

    // Bob decrypts the message with his key.
    const decryptedMessage = encryptDecrypt(encryptedMessage, kb);
    console.log(`Decrypted Message: ${decryptedMessage}`);
}

// Run the main function
main();